### PR TITLE
Enable/disable query quotas per broker

### PR DESCRIPTION
**Pull request by jackjlli**
## Description
This PR adds logic to enable/disable query quotas per Pinot broker. 
This is helpful to test out the capacity of a given broker without affecting other brokers in production. Higher load can be routed to this dedicated host to find out the capacity, which can give us better insights on the current provisioning of the cluster. LinkedIn currently has one service that has been testing in this way. 

The query quotas are all enabled/disabled together within a broker. Once the broker gets restarted, query quotas would all be re-enabled again whatever the previous states were before.

## Tests done:
Started all Pinot components.
Broker log:
```
2020/07/21 15:07:57.813 INFO [HelixExternalViewBasedQueryQuotaManager] [HelixTaskExecutor-message_handle_thread] Initializing rate limiter for table internalTesting_OFFLINE
2020/07/21 15:07:57.813 INFO [HelixExternalViewBasedQueryQuotaManager] [HelixTaskExecutor-message_handle_thread] The number of online brokers for table internalTesting_OFFLINE is 1
2020/07/21 15:07:57.816 INFO [HelixExternalViewBasedQueryQuotaManager] [HelixTaskExecutor-message_handle_thread] Rate limiter for table: internalTesting_OFFLINE has been initialized. Overall rate: 0.01. Per-broker rate: 0.01. Number of online broker instances: 1. Table config stat version: 0
...
2020/07/21 15:08:32.297 INFO [BaseBrokerRequestHandler] [jersey-server-managed-async-executor-5] requestId=6,table=internalTesting_OFFLINE,timeMs=14,docs=20/970,entries=0/480,segments(queried/processed/matched/consuming/unavailable):1/1/1/0/0,consumingFreshnessTimeMs=0,servers=1/1,groupLimitReached=false,exceptions=0,serverStats=(Server=SubmitDelayMs,ResponseDelayMs,ResponseSize,DeserializationTimeMs);172.18.180.84_O=2,9,4515,1,query=select * from internalTesting limit 20
2020/07/21 15:08:33.415 INFO [HelixExternalViewBasedQueryQuotaManager] [jersey-server-managed-async-executor-6] Quota is exceeded for table: internalTesting_OFFLINE. Per-broker rate: 0.01. Current qps: 1
2020/07/21 15:08:33.415 INFO [BaseBrokerRequestHandler] [jersey-server-managed-async-executor-6] Request 7 exceeds query quota for table:internalTesting, query:select * from internalTesting limit 20
2020/07/21 15:08:45.776 INFO [HelixExternalViewBasedQueryQuotaManager] [jersey-server-managed-async-executor-7] Quota is exceeded for table: internalTesting_OFFLINE. Per-broker rate: 0.01. Current qps: 1
2020/07/21 15:08:45.776 INFO [BaseBrokerRequestHandler] [jersey-server-managed-async-executor-7] Request 8 exceeds query quota for table:internalTesting, query:select * from internalTesting limit 30
```
Make the API call to disable query quotas:
`curl 'http://localhost:9000/brokers/instances/Broker_blabla_8099/qps?state=DISABLE' -X 'POST'`
```
2020/07/21 15:08:51.313 INFO [ClusterChangeMediator] [ZkClient-EventThread-16-localhost:2181] Enqueue INSTANCE_CONFIG change
2020/07/21 15:08:51.314 INFO [ClusterChangeMediator] [ClusterChangeHandlingThread] Start processing INSTANCE_CONFIG change
2020/07/21 15:08:51.314 INFO [RoutingManager] [ClusterChangeHandlingThread] Processing instance config change
2020/07/21 15:08:51.335 INFO [RoutingManager] [ClusterChangeHandlingThread] Processed instance config change in 21ms (fetch 3 instance configs: 21ms, calculate changed instances: 0ms) without instance change
2020/07/21 15:08:51.335 INFO [ClusterChangeMediator] [ClusterChangeHandlingThread] Finish handling INSTANCE_CONFIG change for handler: org.apache.pinot.broker.routing.RoutingManager in 21ms
2020/07/21 15:08:51.363 INFO [HelixExternalViewBasedQueryQuotaManager] [ClusterChangeHandlingThread] Set query quota state to: DISABLE for 1 tables in the current broker.
2020/07/21 15:08:51.363 INFO [ClusterChangeMediator] [ClusterChangeHandlingThread] Finish handling INSTANCE_CONFIG change for handler: org.apache.pinot.broker.queryquota.HelixExternalViewBasedQueryQuotaManager in 28ms
2020/07/21 15:08:51.363 INFO [ClusterChangeMediator] [ClusterChangeHandlingThread] Finish processing INSTANCE_CONFIG change in 49ms
2020/07/21 15:08:53.669 INFO [BaseBrokerRequestHandler] [jersey-server-managed-async-executor-7] requestId=9,table=internalTesting_OFFLINE,timeMs=15,docs=30/970,entries=0/720,segments(queried/processed/matched/consuming/unavailable):1/1/1/0/0,consumingFreshnessTimeMs=0,servers=1/1,groupLimitReached=false,exceptions=0,serverStats=(Server=SubmitDelayMs,ResponseDelayMs,ResponseSize,DeserializationTimeMs);172.18.180.84_O=0,11,6275,0,query=select * from internalTesting limit 30
2020/07/21 15:08:57.637 INFO [BaseBrokerRequestHandler] [jersey-server-managed-async-executor-7] requestId=10,table=internalTesting_OFFLINE,timeMs=14,docs=40/970,entries=0/960,segments(queried/processed/matched/consuming/unavailable):1/1/1/0/0,consumingFreshnessTimeMs=0,servers=1/1,groupLimitReached=false,exceptions=0,serverStats=(Server=SubmitDelayMs,ResponseDelayMs,ResponseSize,DeserializationTimeMs);172.18.180.84_O=0,10,8036,0,query=select * from internalTesting limit 40
```
Re-enable query quotas:
`curl 'http://localhost:9000/brokers/instances/Broker_blabla_8099/qps?state=ENABLE' -X 'POST'`
```
2020/07/21 15:09:05.877 INFO [ClusterChangeMediator] [ZkClient-EventThread-16-localhost:2181] Enqueue INSTANCE_CONFIG change
2020/07/21 15:09:05.878 INFO [ClusterChangeMediator] [ClusterChangeHandlingThread] Start processing INSTANCE_CONFIG change
2020/07/21 15:09:05.878 INFO [RoutingManager] [ClusterChangeHandlingThread] Processing instance config change
2020/07/21 15:09:05.898 INFO [RoutingManager] [ClusterChangeHandlingThread] Processed instance config change in 20ms (fetch 3 instance configs: 20ms, calculate changed instances: 0ms) without instance change
2020/07/21 15:09:05.898 INFO [ClusterChangeMediator] [ClusterChangeHandlingThread] Finish handling INSTANCE_CONFIG change for handler: org.apache.pinot.broker.routing.RoutingManager in 20ms
2020/07/21 15:09:05.918 INFO [HelixExternalViewBasedQueryQuotaManager] [ClusterChangeHandlingThread] Set query quota state to: ENABLE for 1 tables in the current broker.
2020/07/21 15:09:05.919 INFO [ClusterChangeMediator] [ClusterChangeHandlingThread] Finish handling INSTANCE_CONFIG change for handler: org.apache.pinot.broker.queryquota.HelixExternalViewBasedQueryQuotaManager in 21ms
2020/07/21 15:09:05.919 INFO [ClusterChangeMediator] [ClusterChangeHandlingThread] Finish processing INSTANCE_CONFIG change in 41ms
2020/07/21 15:09:10.028 INFO [HelixExternalViewBasedQueryQuotaManager] [jersey-server-managed-async-executor-7] Quota is exceeded for table: internalTesting_OFFLINE. Per-broker rate: 0.01. Current qps: 1
2020/07/21 15:09:10.028 INFO [BaseBrokerRequestHandler] [jersey-server-managed-async-executor-7] Request 12 exceeds query quota for table:internalTesting, query:select * from internalTesting limit 50
2020/07/21 15:09:10.920 INFO [HelixExternalViewBasedQueryQuotaManager] [jersey-server-managed-async-executor-7] Quota is exceeded for table: internalTesting_OFFLINE. Per-broker rate: 0.01. Current qps: 2
2020/07/21 15:09:10.920 INFO [BaseBrokerRequestHandler] [jersey-server-managed-async-executor-7] Request 13 exceeds query quota for table:internalTesting, query:select * from internalTesting limit 50
```

### CODE REVIEW

**Comment by mcvsubbu** [comment]
If I understand this right, we are disabling quota check on brokers dynamically. Assuming such dynamic enable/disable is a requirement, why not provide a (rest) API to do it? Why use helix messages?

**Comment by jackjlli** [comment]
> If I understand this right, we are disabling quota check on brokers dynamically. Assuming such dynamic enable/disable is a requirement, why not provide a (rest) API to do it? Why use helix messages?

The reason why I put the API to pinot controller and send a Helix message to broker is that we've already been using this way to update query quotas to brokers (when table config is updated). 
Plus, all the logic of cluster management should all be in one place, which is in Pinot controller.

**Comment by mcvsubbu** [comment]
It will also be confusing if the broker restarts after receiving the message for any reason. It will stop checking for quotas, making debugging harder.
Why not use instance config to set/unset a flag to disable qps quotas?

**Comment by jackjlli** [comment]
> It will also be confusing if the broker restarts after receiving the message for any reason. It will stop checking for quotas, making debugging harder.
> Why not use instance config to set/unset a flag to disable qps quotas?

Updated the PR so that the state gets persisted to broker's instance config

File: pinot-broker/src/main/java/org/apache/pinot/broker/queryquota/HelixExternalViewBasedQueryQuotaManager.java [hunk 1]
```diff
@@ -359,6 +372,27 @@ public void processQueryQuotaChange(ExternalView currentBrokerResource) {
...
+  }
+
+  private void getQueryQuotaEnabledFlagFromInstanceConfig() {
+    Map<String, String> instanceConfigsMap =
+        HelixHelper.getInstanceConfigsMapFor(_instanceId, _helixManager.getClusterName(),
+            _helixManager.getClusterManagmentTool());
+    String queryQuotaEnabled = instanceConfigsMap.getOrDefault(CommonConstants.Helix.QUERY_QUOTA_STATE_ENABLED, "true");
+    _enabled = Boolean.parseBoolean(queryQuotaEnabled);
+    LOGGER.info("Set query quota state to: {} for {} tables in the current broker.", _enabled ? "ENABLE" : "DISABLE",
+        _rateLimiterMap.size());
```
[the remark this answers is not in the document]
Comment (jackjlli) [reply pinot-broker/src/main/java/org/apache/pinot/broker/queryquota/HelixExternalViewBasedQueryQuotaManager.java:389]: I'd insist keeping the size of the map. That'll give us a good insight on how many tables are affected by this state change.
And sure, I can add the state message in the message.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotBrokerRestletResource.java [hunk 2]
```diff
@@ -129,6 +133,40 @@
     }
   }
 
+
+  @POST
+  @Path("/brokers/instances/{instanceName}/qps")
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.TEXT_PLAIN)
+  @ApiOperation(value = "Enable/disable the qps quotas of an broker instance", notes = "Enable/disable the qps quotas of an broker instance")
+  @ApiResponses(value = {@ApiResponse(code = 200, message = "Success"), @ApiResponse(code = 400, message = "Bad Request"), @ApiResponse(code = 404, message = "Instance not found"), @ApiResponse(code = 409, message = "Instance cannot be dropped"), @ApiResponse(code = 500, message = "Internal error")})
```
[the remark this answers is not in the document]
Comment (jackjlli) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotBrokerRestletResource.java:142]: Good catch! 👍 Removed.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotBrokerRestletResource.java [hunk 3]
```diff
@@ -129,6 +133,40 @@
...
+  public SuccessResponse toggleBrokerQpsState(
+      @ApiParam(value = "Broker instance name", required = true, example = "Broker_my.broker.com_30000") @PathParam("instanceName") String brokerInstanceName,
+      @ApiParam(value = "ENABLE|DISABLE", allowableValues = "ENABLE, DISABLE", required = true)  @QueryParam("state") String state) {
+    List<String> liveInstances = _pinotHelixResourceManager.getOnlineInstanceList();
```
[the remark this answers is not in the document]
Comment (jackjlli) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotBrokerRestletResource.java]: Done.

**Comment by jackjlli** [comment]
> Mostly minor comments.
> Two things:
> 
> 1. Should we consider accepting a broker tag name in the API to toggle rate limiting for all brokers in the tag? (as a general functionality, not specific to LinkedIn use case)
> 2. Add a metric (guage) when the config is disabled, so that if it remains disabled for a long time, we can add an alert on it

For 1, I can't think of a scenario for that. But we can always add that later if necessary.
For 2, just added a broker gauge called `QUERY_RATE_LIMIT_DISABLED`.

File: pinot-broker/src/main/java/org/apache/pinot/broker/queryquota/HelixExternalViewBasedQueryQuotaManager.java [hunk 4]
```diff
@@ -352,13 +369,42 @@ public void processQueryQuotaChange(ExternalView currentBrokerResource) {
...
     long endTime = System.currentTimeMillis();
     LOGGER
         .info("Processed query quota change in {}ms, {} out of {} query quota configs rebuilt.", (endTime - startTime),
             numRebuilt, _rateLimiterMap.size());
   }
 
+  /**
+   * Process query quota state change when instance config gets changed
+   */
+  public void processQueryQuotaInstanceConfigChange() {
```
Comment (mcvsubbu) [inline pinot-broker/src/main/java/org/apache/pinot/broker/queryquota/HelixExternalViewBasedQueryQuotaManager.java:385]: suggest renaming these two methods as per new change in config names

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotBrokerRestletResource.java [hunk 5]
```diff
@@ -129,6 +133,40 @@
...
+  @Consumes(MediaType.TEXT_PLAIN)
+  @ApiOperation(value = "Enable/disable the query rate limiting for a broker instance", notes = "Enable/disable the query rate limiting for a broker instance")
+  @ApiResponses(value = {@ApiResponse(code = 200, message = "Success"), @ApiResponse(code = 400, message = "Bad Request"), @ApiResponse(code = 404, message = "Instance not found"), @ApiResponse(code = 500, message = "Internal error")})
+  public SuccessResponse toggleQueryRateLimiting(
+      @ApiParam(value = "Broker instance name", required = true, example = "Broker_my.broker.com_30000") @PathParam("instanceName") String brokerInstanceName,
+      @ApiParam(value = "ENABLE|DISABLE", allowableValues = "ENABLE, DISABLE", required = true)  @QueryParam("state") String state) {
+    if (brokerInstanceName == null || !brokerInstanceName.startsWith("Broker_")) {
+      LOGGER.info("{} isn't a valid broker instance name", brokerInstanceName);
+      throw new ControllerApplicationException(LOGGER,
+          String.format("'%s' is not a valid broker instance name.", brokerInstanceName), Response.Status.BAD_REQUEST);
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotBrokerRestletResource.java:149]: nit: Create an error string, and then use it in log as well as exeption.
Same in other places as well, will not repeat.

Comment (jackjlli) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotBrokerRestletResource.java:148]: Removed the duplicate logging.

**Pull request merged**